cmd/operator-opamp-bridge/internal/proxy/agent.go

```diff
@@ -0,0 +1,170 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package proxy
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"github.com/open-telemetry/opamp-go/server/types"
+	"google.golang.org/protobuf/proto"
+)
+
+type Agent struct {
+	// logger for the agentImpl
+	logger logr.Logger
+	// Agent's instance id. This is an immutable field.
+	InstanceId uuid.UUID
+	// Connection to the Agent.
+	conn types.Connection
+	// mutex for the fields that follow it.
+	mux sync.RWMutex
+
+	health          *protobufs.ComponentHealth
+	effectiveConfig *protobufs.EffectiveConfig
+	// Agent's current status.
+	Status *protobufs.AgentToServer
+}
+
+func NewAgent(logger logr.Logger, agentId uuid.UUID, conn types.Connection) *Agent {
+	return &Agent{
+		logger:     logger,
+		InstanceId: agentId,
+		conn:       conn,
+	}
+}
+
+func (a *Agent) hasCapability(capability protobufs.AgentCapabilities) bool {
+	if capability < 0 {
+		return false
+	}
+	//nolint:gosec
+	return a.Status.Capabilities&uint64(capability) != 0
+}
+
+func (a *Agent) updateAgentDescription(newStatus *protobufs.AgentToServer) (agentDescrChanged bool) {
+	prevStatus := a.Status
+
+	if a.Status == nil {
+		// First time this Agent reports a status, remember it.
+		a.Status = newStatus
+		agentDescrChanged = true
+	} else {
+		// Not a new Agent. Update the Status.
+		a.Status.SequenceNum = newStatus.SequenceNum
+
+		// Check what's changed in the AgentDescription.
+		if newStatus.AgentDescription != nil {
+			// If the AgentDescription field is set it means the Agent tells us
+			// something is changed in the field since the last status report
+			// (or this is the first report).
+			// Make full comparison of previous and new descriptions to see if it
+			// really is different.
+			if prevStatus != nil && proto.Equal(prevStatus.AgentDescription, newStatus.AgentDescription) {
+				// Agent description didn't change.
+				agentDescrChanged = false
+			} else {
+				// Yes, the description is different, update it.
+				a.Status.AgentDescription = newStatus.AgentDescription
+				agentDescrChanged = true
+			}
+		} else {
+			// AgentDescription field is not set, which means description didn't change.
+			agentDescrChanged = false
+		}
+
+		// Update remote config status if it is included and is different from what we have.
+		if newStatus.RemoteConfigStatus != nil &&
+			!proto.Equal(a.Status.RemoteConfigStatus, newStatus.RemoteConfigStatus) {
+			a.Status.RemoteConfigStatus = newStatus.RemoteConfigStatus
+		}
```

Comment:
This also doesn't feel like it belongs in a function which is only supposed to update the description.

Reply:
yeah a lot of this is boilerplate from other server implementations. Re-organizing now.

Reply:
i can actually get rid of this, because i'm not supporting remote config through the bridge proxy right now. That'll be future improvements.